KAFKA-13015 Ducktape System Tests for Metadata Snapshots

niket-goel opened the pull request:
This PR implements system tests in ducktape to test the ability of brokers and controllers to generate and consume snapshots and catch up with the metadata log.

Test Output:
```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.8.1
session_id:       2021-07-15--017
run time:         6 minutes 46.061 seconds
tests run:        2
passed:           2
failed:           0
ignored:          0
================================================================================
test_id:    kafkatest.tests.core.snapshot_test.TestSnapshots.test_broker.metadata_quorum=REMOTE_KRAFT
status:     PASS
run time:   4 minutes 3.388 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.snapshot_test.TestSnapshots.test_controller.metadata_quorum=REMOTE_KRAFT
status:     PASS
run time:   2 minutes 42.434 seconds
--------------------------------------------------------------------------------

```

niket-goel commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,227 @@
+            self.logger.debug("File %s was found"%file_path)
+            return True
+
+    def validate_success(self):
```
These are not the best checks IMO, I just borrowed them from some other test. Please let me know If there are better checks that we could perform.

jsancio commented on tests/kafkatest/services/kafka/kafka.py:
```diff
@@ -154,6 +154,9 @@ class for details.
     CONFIG_FILE = os.path.join(PERSISTENT_ROOT, "kafka.properties")
+    METADATA_LOG_DIR = os.path.join (PERSISTENT_ROOT, "kafka-metadata-logs")
+    METADATA_SNAPSHOT_SEARCH_STR = "%s/@metadata-0/*.checkpoint" % METADATA_LOG_DIR
+    METADATA_FIRST_LOG = "%s/@metadata-0/00000000000000000000.log" % METADATA_LOG_DIR
```
Should we use `os.path.join` for these two path manipulations?

jsancio commented on tests/kafkatest/services/kafka/templates/log4j.properties:
```diff
@@ -111,6 +111,8 @@ log4j.logger.kafka.client.ClientUtils={{ log_level|default("DEBUG") }}, kafkaInf
 log4j.logger.kafka.perf.ProducerPerformance$ProducerThread={{ log_level|default("DEBUG") }}, kafkaInfoAppender, kafkaDebugAppender
 log4j.logger.kafka={{ log_level|default("DEBUG") }}, kafkaInfoAppender, kafkaDebugAppender
+log4j.logger.org.apache.kafka.raft={{ log_level|default("DEBUG") }}, kafkaInfoAppender, kafkaDebugAppender
+log4j.logger.org.apache.kafka.snapshot={{ log_level|default("DEBUG") }}, kafkaInfoAppender, kafkaDebugAppender
```
I think we can remove these lines since the root logger is already configured to this same level to stdout:
```
log4j.rootLogger={{ log_level|default("DEBUG") }}, stdout
```

niket-goel replied:
Ack. Left in by mistake.

jsancio commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,227 @@
+    def create_n_topics(self, topic_count):
+        for i in range(self.topics_created, topic_count):
+            topic = "test_topic_%d" % i
+            print("Creating topic %s" % topic, flush=True)
```
Use `self.logger`.

jsancio commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,227 @@
+    def wait_for_log_segment_delete(self, node):
+        file_path = self.kafka.METADATA_FIRST_LOG
+        # Wait until the first log segment in metadata log is marked for deletion
+        wait_until(lambda: self.file_exists(node, file_path) == False,
```
`lambda: not self.file_exists(node, file_path)`

jsancio commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,227 @@
+    def wait_for_snapshot(self, node):
+        # Wait for a snapshot file to show up
+        file_path = self.kafka.METADATA_SNAPSHOT_SEARCH_STR
+        wait_until(lambda: self.file_exists(node, file_path) == True,
```
`lambda: self.file_exists(node, file_path)`

jsancio commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,227 @@
+        if self.kafka.remote_controller_quorum:
+            controller_nodes = self.kafka.remote_controller_quorum.nodes
+        else:
+            controller_nodes = self.kafka.nodes[:self.kafka.num_nodes_controller_role]
```
Should we move this complexity of remote vs co-located to the Kafka service?

niket-goel replied:
We could do that. It would just be a bigger change to get expose controller nodes correctly through state transitions/modifications within the service (I think). I would maybe not do it in this PR?

jsancio commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,227 @@
+        else:
+            # This is colocated kraft
+            # Picking the first node as that is guaranteed to be a controller
+            controller_node = self.kafka.nodes[0]
```
Should we also make this random?

niket-goel replied:
Yes.. We can make it a random selection as well. The reason I left it this way was for simplicity (colocated mode has both controller and non controller nodes in the same list)

jsancio commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,227 @@
+
+    def file_exists(self, node, file_path):
+        # Check if the first log segment is cleaned up
+        self.logger.debug("Checking if file %s exists"%file_path)
```
Missing spaces before and after the `%` string formatter. This applies to a few places I this file.

jsancio commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,227 @@
+            self.logger.debug("File %s was found"%file_path)
+            return True
+
+    def validate_success(self):
+        cluster_id = self.kafka.cluster_id()
+        assert cluster_id is not None
+        assert len(cluster_id) == 22
+        assert self.kafka.all_nodes_support_topic_ids()
+        assert self.kafka.check_protocol_errors(self)
```
Is this enough to verify success?

jsancio replied:
I took a look at the implementations for `self.kafka.all_nodes_support_topic_ids` and `self.kafka.check_protocol_errors` and it doesn't look like these checks are enough.

Also, this call `self.kafka.check_protocol_errors(self)` seems strange. It looks like the implementation of that method never uses that parameter.

niket-goel replied:
Same comment as the one on the controller success validation. This is not a good verification method at the moment, I just borrowed it from another test. For us I think the fact that the nodes are up and we can produce/consume should be a good check. I will make a change to have that as the verification.

jsancio commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,227 @@
+        # Create a topic where the affected broker must be the leader
+        broker_topic = "test_topic_%d" % self.topics_created
+        self.topics_created += 1
+        print("Creating topic %s" % broker_topic, flush=True)
```
Use `self.logger`.

niket-goel replied:
So the reason I used this is because other tests force the output of topic create to stdout (for Jenkins I think).I can move this to the logger instead.

jsancio commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,227 @@
+
+    @cluster(num_nodes=9)
+    @parametrize(metadata_quorum=quorum.remote_kraft)
+    # TODO: enable after concurrent startup has been merged
```
Are you planning to implement this TODO in this PR? If not, let's remove this TODO and create a Jira.

Do we need to fix this issue before 3.0? In other words is this a blocker for 3.0?

niket-goel replied:
I wanted to discuss this with you. Let me ping offline about this. Basically we are not testing colocated craft at this point.

jsancio commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,227 @@
+        # Scenario -- Re-init broker after cleaning up all persistent state
+        # and create a topic where the affected broker must be the leader
+        node = random.choice(self.kafka.nodes)
+        self.logger.debug("Scenario 3: kill-clean-start-verify-produce on node %s", self.kafka.who_am_i(node))
```
Do we think that we need these 3 scenarios in this test? Do you think that we only need this scenario to verify that snapshot are getting generated and loaded by the brokers?

niket-goel replied:
I agree that the third scenario covers the initial cases as well, but the reason I left the other ones in was to make debugging failures easier. If a subset of the scenarios work, it can give help investigate where the failure might be. The scenarios themselves take 10s of seconds to execute. Do you think leaving them in is helpful?

jsancio commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,227 @@
+        self.logger.debug("Scenario 2: kill-clean-start on node %s", self.kafka.who_am_i(node))
+        self.kafka.controller_quorum.clean_node(node)
+        self.kafka.controller_quorum.start_node(node)
+        self.validate_success()
```
It is not clear to me that this is enough to verify that the controller loaded a snapshot and caught to the leader.

niket-goel replied:
I agree that this is not a good check. I was trying to look for verification that other tests do, but it seems that maybe topic creation and production/consumption is the best verification. I will do that here as well.

cmccabe replied:
I agree with @jsancio . What about if we bring down all the controller nodes and then bring them back? They'd have to load snapshot in that case, right, or else the quorum would not work?

niket-goel replied:
Bringing down any one controller, cleaning up it's kafka dir and then rebooting is already ensuring that the snapshot is being loaded on recovery. I think Jose's comment was on the previous version of the the test, where we were not doing a produce/consume check. In this version of the PR we are creating a topic and verifying we can produce to it and consume from it. The rationale is that to be able to create topic and produce to it, the quorum needs to agree on the metadata up until this point.

I take your point that maybe when testing the controller, it would be a better check to bring all controllers down, restart them and then do a produce/consume check as that would ensure that a quorum of nodes was able to load from their snapshots and continue operation. Let me make that change. Will update the PR.

niket-goel replied:
Updated to bounce all controllers in the latest revision.

jsancio replied:
> What about if we bring down all the controller nodes and then bring them back?

Yes. Here are the steps in a bit more details:
1. Check that all of the controllers deleted their log segment starting at base offset 0.
2. For each controller one at a time:
  a. shutdown the controller
  b. delete all of the local data
  c. start the controller
  d. wait until it joins the quorum. I am not sure we have the tools to do this today. Maybe for now we need to sleep and file a Jira to build this tooling.
3. Create a new topic, produce to it, consume from it. This shows that the quorum is available and the brokers are fetching from it.

niket-goel replied:
Agree with you @jsancio . The latest commit for the PR should already be doing this.

jsancio reviewed (commented):
Thanks for the PR @niket-goel

niket-goel commented:
The failure seen on Jenkins [1] does not reproduce locally. I am also seeing the same failure on a different unrelated PR [2]

[1]
```
[2021-07-20T17:32:12.029Z] ConsumerBounceTest > testCloseDuringRebalance() FAILED

[2021-07-20T17:32:12.029Z]     org.opentest4j.AssertionFailedError: Rebalance did not complete in time ==> expected: <true> but was: <false>
```

[2] https://github.com/apache/kafka/pull/11070

cmccabe commented on tests/kafkatest/services/kafka/config.py:
```diff
@@ -24,6 +24,11 @@ class KafkaConfig(dict):
         config_property.LOG_DIRS: "/mnt/kafka/kafka-data-logs-1,/mnt/kafka/kafka-data-logs-2",
+        config_property.METADATA_LOG_DIR: "/mnt/kafka/kafka-metadata-logs",
+        config_property.METADATA_LOG_SEGMENT_BYTES: 8388608,
+        config_property.METADATA_LOG_BYTES_BETWEEN_SNAPSHOTS: 100,
```
Hmm, I'm a bit confused about why we'd make something like "100 bytes between snapshots" the default. I guess the intention is to do a stress test?

niket-goel replied:
The idea was to initially stress test it. Then we ended up choosing a higher value for this config which ends up overriding this value. Will fix the default setting here to be higher.

cmccabe commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,224 @@
+
+class TestSnapshots(ProduceConsumeValidateTest):
+
+    METADATA_LOG_RETENTION_BYTES = "2048"
```
I think you can just put the constants later on in the `server_prop_overrides` block. There's no reason to have a special name for them since they're only used in one place (which is itself a constant)

niket-goel replied:
Fair point. Will remove these definitions.

cmccabe commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,224 @@
+        # Each scenario is a progression over the previous one.
+        # Redundant scenarios are left in to make debugging a failure of the test easier
+
+        # Scenario -- Re-init broker after cleaning up all persistent state
```
Should this be a separate test function? Putting it here seems like it might muddy the waters, especially if the nodes you hit are combined controller/brokers.

niket-goel replied:
I am not sure I got your point. Will sync offline about this.

niket-goel replied:
Synced offline. Breaking away into separate test functions would greatly increase the test runtime. Siding towards keeping them in one function. The reasoning for multiple scenarios is now added to the test description

niket-goel committed:
KAFKA-13015: Ducktape System Tests to verify brokers' and controllers' ability to create
metadata snapshots and load them to catchup to the latest state of the cluster metadata

niket-goel committed:
Changing some defaults and changing the metadata topic name

niket-goel commented:
Rebased the branch with Trunk

niket-goel committed:
Removing a redundant scenario from test_controller

jsancio commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,251 @@
+        Even though a subsequent scenario covers the previous one, they are all
+        left in the test to make debugging a failure of the test easier
+        e.g. if the first scenario passes and the second fails, it hints towards
+        a problem with the application of delta records while catching up
```
For this to be true don't we need to call `validate_success` after every scenario?

niket-goel replied:
What you are suggesting is definitely a stronger check. The reason I did not do that was :
  1. The way the code is laid out, doing a produce consume cycle needs two additional nodes each time. Doing it multiple times would have needed (num_scenarios*2) extra nodes to run the test.
  2. Being able to start up the controller/broker without error is a check by itself (we caught bugs just while starting up/creating topics), and for completeness the final validate_success validates all prior actions.

I can address this concern the following ways:
  1. I could modify code to have the same Producer/Consumer node do validation for multiple topics, but that might take time.
  2. I could add more nodes to the test to account for the extra producers/consumers.
  3. Check in the code as is for now and improve it (alongside other test cases in another iteration)

What would you prefer?

jsancio commented on tests/kafkatest/tests/core/snapshot_test.py:
```diff
@@ -0,0 +1,251 @@
+        Even though a subsequent scenario covers the previous one, they are all
+        left in the test to make debugging a failure of the test easier
+        e.g. if the first scenario passes and the second fails, it hints towards
+        a problem with the application of delta records while catching up
```
For this to be true don't we need to call validate_success after every scenario?

niket-goel replied:
Same as below

cmccabe reviewed (approved):
LGTM.

Let's get this in and we can work on tightening up the tests a bit post-3.0.
